### PR TITLE
Cmplog instrumentation for Frida

### PR DESCRIPTION
**Pull request by OmreeBenari**
add cmplog runtime instrumentation

### CODE REVIEW

**Comment by OmreeBenari** [comment]
For some reason I can't reproduce Clippy's errors on my environment.. 
besides few warning Clippy finishes successfully

File: Cargo.toml [hunk 1]
```diff
@@ -11,13 +11,14 @@ members = [
     "libafl_cc",
     "libafl_targets",
     "libafl_frida",
+    "fuzzers/frida_libpng",
```
Comment (andreafioraldi) [inline Cargo.toml:14]: This file must not be changed

File: fuzzers/frida_libpng/harness.cc [hunk 2]
```diff
@@ -96,19 +96,32 @@ void func3( char * alloc) {
...
+    //printf("func1\n");
   func2();
 }
 // Entry point for LibFuzzer.
 // Roughly follows the libpng book example:
 // http://www.libpng.org/pub/png/book/chapter13.html
 extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
-  if (size < kPngHeaderSize) {
+
+   if(data[10] == 0xAB &&
```
Comment (andreafioraldi) [inline fuzzers/frida_libpng/harness.cc:112]: This test cannot be passed as now there isn't a stage using CmpLog in the lib

File: libafl_frida/Cargo.toml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 [package]
 name = "libafl_frida"
 version = "0.2.0"
-authors = ["s1341 <github@shmarya.net>"]
+authors = ["s1341 <github@shmarya.net>", "OmreeBenari <github@Omreeb.net>"]
```
Comment (andreafioraldi) [inline libafl_frida/Cargo.toml:4]: This fields is for the author, not contributors. @domenukk may we should add a contributors section in the readme, like the one in aflpp.

File: libafl_frida/src/cmplog_rt.rs [hunk 4]
```diff
@@ -0,0 +1,187 @@
+use dynasmrt::{dynasm, DynasmApi, DynasmLabelApi};
+use nix::{
+    libc::{memmove, memset},
+    sys::mman::{mmap, MapFlags, ProtFlags},
+};
+use std::ffi::c_void;
+
+// TODO compile time flag
+//pub const CMPLOG_MAP_W: usize = 65536;
+pub const CMPLOG_MAP_W: usize = 8000;
```
Comment (andreafioraldi) [inline libafl_frida/src/cmplog_rt.rs:10]: This code is duplicated from libafl_targets. Btw, I reworked CmpLog in branch (cmplog branch), there you can find directly some function that you can call, don't rely on sancov functions for frida

**Comment by andreafioraldi** [comment]
Hi, thank you for the contribution. ATM we cannot test it due to the lack of a Mutator, it is in the roadmap.
Please rebase it to the cmplog branch. This PR cannot be merged until I implement at leat one Mutator using CmpLog as I will likely hange the API.

File: libafl_frida/Cargo.toml (hunk 3 above)
Comment (domenukk) [reply libafl_frida/Cargo.toml:4]: I don't think we need a specific contributors section, it's always more out of date and less accurate than the github contributors section, right?
WRT author, @s1341 and @OmreeBenari can figure out what seems right in this case.

File: fuzzers/frida_libpng/harness.cc (hunk 2 above)
Comment (OmreeBenari) [reply fuzzers/frida_libpng/harness.cc:112]: should i delete it entirely or keep it as a comment?

**Comment by andreafioraldi** [comment]
Hi @OmreeBenari , a first naive mutator based on cmplog is now on main, if you want you can rebase your changes on main and adapt your code to work with the new libafl_targets interface for cmplog

File: libafl_frida/src/cmplog_rt.rs (hunk 4 above)
Comment (OmreeBenari) [reply libafl_frida/src/cmplog_rt.rs:10]: did you meant __libafl_targets_cmplog?

Comment (andreafioraldi) [reply libafl_frida/src/cmplog_rt.rs:10]: You can use directly libafl_targets::cmplog (https://github.com/AFLplusplus/LibAFL/blob/main/libafl_targets/src/cmplog.rs). For the instrumentation routine, yes it is __libafl_target_cmplog and it is in C (https://github.com/AFLplusplus/LibAFL/blob/main/libafl_targets/src/cmplog.h). To expose it to rust, you can simply create a non static stub that call this static function and use it from Rust as extern "C"

**Comment by domenukk** [comment]
Hey, what's missing here? Looks like the clippy issue in ci already got fixed in `main`, anything else we can help out with?

**Comment by OmreeBenari** [comment]
> Hey, what's missing here? Looks like the clippy issue in ci already got fixed in `main`, anything else we can help out with?

Hey, sorry I missed your answer.

I've reabsed my code on main again in case I've missed  the relevant commit. @andreafioraldi  did you look into the observer?

either way it seems like there's some more work to do, it's currently panicking and I need to look into it(it did not panic before rebasing when the cmplog observer was disabled)

**Review by s1341 — APPROVED** [review]
Good work!

**Review by andreafioraldi — APPROVED** [review]

**Pull request merged**